extensions/amp-instagram/1.0/instagram.js

```diff
@@ -0,0 +1,89 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {ContainWrapper} from '../../../src/preact/component';
+import {getData} from '../../../src/event-helper';
+import {useMountEffect} from '../../../src/preact/utils';
+import {useState} from '../../../src/preact';
+
+/**
+ * @param {!InstagramProps} props
+ * @return {PreactDef.Renderable}
+ */
+export function Instagram({shortcode, captioned, style, alt, resize}) {
+  const iframeRef = Preact.useRef(null);
+  const [heightStyle, setHeightStyle] = useState(null);
+  const [opacity, setOpacity] = useState(0);
+
+  /**
+   * Upon component mount, the Instagram post's height is received and applied.
+   */
+  useMountEffect(() => {
+    /**
+     * @param {Event} event
+     */
+    function handleMessage(event) {
+      if (
+        event.origin != 'https://www.instagram.com' ||
+        event.source != iframeRef.current.contentWindow
+      ) {
+        return;
+      }
+
+      const data = JSON.parse(getData(event));
+
+      if (data['type'] == 'MEASURE') {
+        if (typeof resize === 'function') {
+          resize(data['details']['height']);
+        } else {
+          setHeightStyle({'height': data['details']['height']});
+        }
+        setOpacity(1);
+      }
+    }
+
+    window.addEventListener('message', handleMessage);
+
+    return () => {
+      window.removeEventListener('message', handleMessage);
+    };
+  });
+
+  return (
+    <ContainWrapper style={{...style, ...heightStyle}} layout size paint>
+      <iframe
+        ref={iframeRef}
+        src={
+          'https://www.instagram.com/p/' +
+          (encodeURIComponent(shortcode) || 'error') +
```

Comment:
Here, if the shortcode isn't passed in at all, then we change it to 'error' so that the URL still is valid. That way the user is displayed the embedded Instagram error iframe, which is a clearer error message + cleaner look than a custom error handling implementation.

Reply:
Hmm. I think running an iframe that will knowingly fail is not a good idea. I'd rather not render an iframe at all. But my question is: is `!encodeURIComponent(shortcode)` a valid use case? Or is it some sort of typing violation. If it's a typing violation, I'd rather not overmassage this case and let errors be.